addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
         }
     }
 
-    private void fadeToState(final InternalLedState newState) throws IOException {
-        if (!newState.equals(realTargetState)) {
-            keepFading = false;
-            realTargetState = newState;
+    final static class LEDFaderRunner implements Runnable {
+        protected Logger logger = LoggerFactory.getLogger(LEDFaderRunner.class);
+        InternalLedState fadeState;
+        int fadeStepper = 1;
+
+        final InternalLedState toState;
+        final InternalLedState fromState;
+        final Future<?>[] faderArray;
+        final int steps;
+        final Function<InternalLedState, Boolean> ledSender;
+
+        public LEDFaderRunner(Future<?>[] faderFuture, InternalLedState fromState, InternalLedState toState, int steps, Function<InternalLedState,  Boolean> ledSender) {
+            this.faderArray = faderFuture;
+            fadeState=fromState;
+            this.fromState = fromState;
+            this.toState = toState;
+            this.steps = steps;
+            this.ledSender = ledSender;
+        }
 
-            executorService.schedule(() -> {
-                if (currentTargetState.equals(newState)) {
-                    return;
+        public void run() {
+            Future<?> thisFuture;
+            while (null == (thisFuture = faderArray[0])) Thread.yield();
+            if (fadeStepper <= steps) {
+                if (!Thread.currentThread().isInterrupted()) {
+                    fadeState = fromState.fade(toState, (double) fadeStepper / steps);
+                    logger.debug("fadeState: {}", fadeState);
+                    if (!ledSender.apply(fadeState)) {
+                        logger.debug("fadeState: sending failed at step {}", fadeStepper);
+                        thisFuture.cancel(true);
```

Comment:
This means there was an error fading, maybe it is better to leverage the default functionality of the scheduler to cancel the job as soon as an error is thrown, so just throw some (specific) RuntimeException and it will cancel no need for any further hacks imho.